### PR TITLE
ENG-6511 mview minmax index optimization

### PR DESCRIPTION
**Pull request by yzhang1991**

### CODE REVIEW

File: src/ee/storage/MaterializedViewMetadata.cpp [hunk 1]
```diff
@@ -298,29 +313,82 @@ NValue MaterializedViewMetadata::findMinMaxFallbackValueIndexed(const TableTuple
...
+                break;
+            }
+            // skip the oldTuple and apply post filter
+            if (tuple.equals(oldTuple) ||
+                (m_filterPredicate && !m_filterPredicate->eval(&tuple, NULL).isTrue())) {
+                continue;
+            }
+            NValue current = (aggExpr) ? aggExpr->eval(&tuple, NULL) : tuple.getNValue(srcColIdx);
+            if (current.isNull()) {
+                continue;
```
Comment (paulmartel) [inline src/ee/storage/MaterializedViewMetadata.cpp]: Can't we return initialNull as soon as we detect current.isNull()?

File: src/ee/storage/MaterializedViewMetadata.cpp [hunk 2]
```diff
@@ -298,29 +313,82 @@ NValue MaterializedViewMetadata::findMinMaxFallbackValueIndexed(const TableTuple
...
+            m_indexForMinMax->moveToGreaterThanKey(&m_minMaxSearchKeyTuple, minMaxCursor);
+            m_indexForMinMax->moveToPriorEntry(minMaxCursor);
+        }
+        while ( ! (tuple = m_indexForMinMax->nextValue(minMaxCursor)).isNullTuple() ) {
+            // If the cursor already moved out of the target group range, exit the loop.
+            bool matchGroupBy = true;
+            for (int colindex = 0; colindex < m_groupByColumnCount; colindex++) {
+                NValue value = getGroupByValueFromSrcTuple(colindex, tuple);
+                if ( value.compare(m_minMaxSearchKeyValue[colindex]) != 0 ) {
+                    matchGroupBy = false;
```
Comment (paulmartel) [inline src/ee/storage/MaterializedViewMetadata.cpp]: Remove the matchGroupBy flag and immediately return initialNull on a mismatch.

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 3]
```diff
@@ -2104,6 +2104,8 @@ void processMaterializedViews(Database db) throws VoltCompiler.VoltCompilerExcep
             List<AbstractExpression> aggregationExprs = new ArrayList<AbstractExpression>();
             boolean hasAggregationExprs = false;
             boolean hasMinOrMaxAgg = false;
+            boolean hasOnlyOneDistinctMinOrMaxAggExpr = false;
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Don't use the word "Distinct" -- it is confusing, given the technical meaning of "distinct" in sql.

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 4]
```diff
@@ -2177,43 +2185,30 @@ void processMaterializedViews(Database db) throws VoltCompiler.VoltCompilerExcep
...
+            Table srcTable, List<AbstractExpression> groupbyExprs, AbstractExpression singleDistinctMinMaxAggExpr)
     {
         CatalogMap<Index> allIndexes = srcTable.getIndexes();
         StmtTableScan tableScan = new StmtTargetTableScan(srcTable, srcTable.getTypeName());
 
+        // Candidate index. If we can find an index covering both group-by columns and aggExpr (optimal) then we will
+        // return immediately. If the index found covers only group-by columns (sub-optimal), we will first cache it here.
+        Index candidate = null;
         for (Index index : allIndexes) {
+            // matchedAll == ture if the index covered all group-by columns (sub-optimal candidate).
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: spelling true

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 5]
```diff
@@ -2132,10 +2140,10 @@ void processMaterializedViews(Database db) throws VoltCompiler.VoltCompilerExcep
...
 
             if (hasMinOrMaxAgg) {
-                // TODO: deal with minMaxAggs, i.e. if only one min/max agg, try to find the index
-                // with group by cols followed by this agg col; if multiple min/max aggs, decide
-                // what to do (probably the index on group by cols is the best choice)
-                Index found = findBestMatchIndexForMatviewMinOrMax(matviewinfo, srcTable, groupbyExprs);
+                // ENG-6511: If we have only one distinct min/max aggCol/aggExpr, we will pass it into
+                // findBestMatchIndexForMatviewMinOrMax() to see if a better index can be found.
+                AbstractExpression singleDistinctMinMaxAggExpr = hasOnlyOneDistinctMinOrMaxAggExpr ? minMaxAggs.get(0) : null;
+                Index found = findBestMatchIndexForMatviewMinOrMax(matviewinfo, srcTable, groupbyExprs, singleDistinctMinMaxAggExpr);
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: we don't need a boolean to tell that we found the expression -- just cache the expression (or null).

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 6]
```diff
@@ -2116,6 +2118,12 @@ void processMaterializedViews(Database db) throws VoltCompiler.VoltCompilerExcep
                         col.expression.getExpressionType() == ExpressionType.AGGREGATE_MAX) {
                     hasMinOrMaxAgg = true;
                     minMaxAggs.add(aggExpr);
+                    /*
+                     * ENG-6511: If we only have one distinct min/max aggCol/aggExpr, we can try to
+                     * find an index that was built on both the group-by columns and the min/max
+                     * aggCol/aggExpr to achieve better performance.
+                     */
+                    hasOnlyOneDistinctMinOrMaxAggExpr = aggExpr.equals(minMaxAggs.get(0));
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: bug?    AS SELECT G, COUNT(*), MAX(A), MAX(B), MIN(A) FROM ...

Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: test this case.

Comment (yzhang1991) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Yes this is a bug.

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 7]
```diff
@@ -2221,17 +2222,131 @@ private static Index findBestMatchIndexForMatviewMinOrMax(MaterializedViewInfo m
...
+                    CatalogUtil.getSortedCatalogItems(matviewinfo.getGroupbycols(), "index");
+                if (indexedExprs == null) { // groupbyExprs == null && indexedExprs == null
+                    // All the columns in the index are also simple columns, EASY! colref vs. colref
+                    List<ColumnRef> indexedColRefs =
+                        CatalogUtil.getSortedCatalogItems(index.getColumns(), "index");
+                    // The number of columns in index can never be less than that in the group-by column list.
+                    // If singleUniqueMinMaxAggExpr == null, they must be equal (diffAllowance == 0)
+                    // Otherwise they may be equal (sub-optimal) or
+                    // indexedColRefs.size() == groupbyColRefs.size() + 1 (optimal, diffAllowance == 1)
+                    if ( indexedColRefs.size() < groupbyColRefs.size() ||
```
Comment (jxchinaren) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Can you refactor this Index size IF check into a function? It has been used 6 times in this block.

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 8]
```diff
@@ -2221,17 +2222,131 @@ private static Index findBestMatchIndexForMatviewMinOrMax(MaterializedViewInfo m
...
+                         indexedColRefs.size() > groupbyColRefs.size() + diffAllowance ) {
+                        continue;
+                    }
+                    // Compare group by columns.
```
Comment (jxchinaren) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Also, can you refactor this "Compare group by columns/expressions" logic into a function? The similar logic has been used various times.

**Comment by jxchinaren** [comment]
I am done reviewing the DDLCompiler.java, not sure what's left for review, will coordinate with Paul about that.

File: src/frontend/org/voltdb/compiler/DDLCompiler.java (hunk 8 above)
Comment (yzhang1991) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: It is hard to do so, there are 3 group-by column comparison here but they each have unique type conversion / comparison. If we refactor those logic out, I don't see how we can make the code simpler. What do you think?

Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: @jxchinaren Xin, can you suggest specific function names/signatures for the refactored functions you are suggesting?

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 9]
```diff
@@ -2177,43 +2191,30 @@ void processMaterializedViews(Database db) throws VoltCompiler.VoltCompilerExcep
...
-                String expressionjson = index.getExpressionsjson();
-                if (expressionjson.isEmpty()) {
-                    continue;
-                }
-                List<AbstractExpression> indexedExprs = null;
+            // optimal == true if the index covered both the group-by columns and the min/max aggExpr.
+            boolean optimal = false;
+            // If singleUniqueMinMaxAggExpr is not null, the diff can be zero or one.
+            // Otherwise, for a usable index, its number of columns must agree with that of the group-by columns.
+            int diffAllowance = singleUniqueMinMaxAggExpr == null ? 0 : 1;
```
Comment (jxchinaren) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: make "diffAllowance" final?

**Pull request merged**